### PR TITLE
fix(zkevm_api): getBatchByNumber, ler, other fixes

### PR DESCRIPTION
**Pull request by revitteth**
## Sequencer
- read s/c once per batch to get/store the LER into hermezdb table

## RPC
- stores LER from stream against batch number
- if not found in db on zkevm_api JSON RPC call, retrieves from S/C storage

### CODE REVIEW

File: cmd/rpcdaemon/commands/zkevm_api.go [hunk 1]
```diff
@@ -357,7 +358,12 @@ func (api *ZkEvmAPIImpl) GetBatchByNumber(ctx context.Context, batchNumber rpc.B
 	if seq != nil {
 		batch.SendSequencesTxHash = &seq.L1TxHash
 	}
-	batch.Closed = seq != nil || bn == 0 || bn == 1 // sequenced, genesis or injected batch 1 - special batches 0,1 will always be closed
+	_, found, err := hermezDb.GetLowestBlockInBatch(bn + 1)
```
Comment (hexoscott) [inline cmd/rpcdaemon/commands/zkevm_api.go]: could we rename `bn` to batchNumber or similar, bn could mean block number or batch number and the context gets confusing

File: cmd/rpcdaemon/commands/zkevm_api.go [hunk 2]
```diff
@@ -368,7 +374,7 @@ func (api *ZkEvmAPIImpl) GetBatchByNumber(ctx context.Context, batchNumber rpc.B
 		batch.VerifyBatchTxHash = &ver.L1TxHash
 	}
 
-	// exit roots
+	// exit roots (mer, rer)
```
Comment (hexoscott) [inline cmd/rpcdaemon/commands/zkevm_api.go]: could we use the long form rather than abbreviations

File: zk/utils/utils.go [hunk 3]
```diff
@@ -123,3 +125,43 @@ func RecoverySetBlockConfigForks(blockNum uint64, forkId uint64, cfg ForkConfigW
...
+		checkBatch := batchNo
+		for ; checkBatch > 0; checkBatch-- {
+			blockNo, err := db.GetHighestBlockInBatch(checkBatch)
+			if err != nil {
+				return libcommon.Hash{}, err
+			}
+			stateReader := state.NewPlainState(tx, blockNo, nil)
+			rawLer, err := stateReader.ReadAccountStorage(state.GER_MANAGER_ADDRESS, 1, &state.GLOBAL_EXIT_ROOT_POS_1)
+			if err != nil {
+				return libcommon.Hash{}, err
```
Comment (hexoscott) [inline zk/utils/utils.go:157]: maybe we need to also a `stateReader.Close()` line here a well

**Review by hexoscott — CHANGES_REQUESTED** [review]
some tiny comments but looks good to me

**Review by hexoscott — DISMISSED** [review]

File: cmd/rpcdaemon/commands/zkevm_api.go [hunk 4]
```diff
@@ -230,6 +230,130 @@ func (api *ZkEvmAPIImpl) VerifiedBatchNumber(ctx context.Context) (hexutil.Uint6
...
+		}
+
+		// block numbers in batch
+		blocksInBatch, err := hermezDb.GetL2BlockNosByBatch(batchNo)
+		if err != nil {
+			return nil, err
+		}
+
+		// collect blocks in batch
+		var batchBlocks []*eritypes.Block
```
Comment (hexoscott) [inline cmd/rpcdaemon/commands/zkevm_api.go:285]: this seems a repeated block of code, can it brought into its own function?

**Review by hexoscott — CHANGES_REQUESTED** [review]

File: cmd/rpcdaemon/commands/zkevm_api.go (hunk 4 above)
Comment (revitteth) [reply cmd/rpcdaemon/commands/zkevm_api.go:285]: todo'd this one :)

File: cmd/rpcdaemon/commands/zkevm_api.go [hunk 5]
```diff
@@ -308,19 +437,19 @@ func (api *ZkEvmAPIImpl) GetBatchByNumber(ctx context.Context, batchNumber rpc.B
...
 	// global exit root of batch
-	batchGer, foundBatchGerNumber, err := hermezDb.GetLastBatchGlobalExitRoot(bn)
+	batchGer, foundBatchGerNumber, err := hermezDb.GetLastBatchGlobalExitRoot(batchNo)
 	if err != nil {
 		return nil, err
 	}
 
 	// get last block in batch
-	lastBlockInbatch, err := hermezDb.GetHighestBlockInBatch(bn)
+	lastBlockInbatch, err := hermezDb.GetHighestBlockInBatch(batchNo)
```
Comment (hexoscott) [inline cmd/rpcdaemon/commands/zkevm_api.go:452]: do we not already do this on line 391?

**Review by hexoscott — DISMISSED** [review]
one small comment, but looks good

**Review by hexoscott — APPROVED** [review]

**Pull request merged**